kernel_gateway/services/kernels/manager.py

```diff
@@ -63,3 +65,108 @@ def start_kernel(self, *args, **kwargs):
                         self.shutdown_kernel(kernel_id)
                         raise RuntimeError('Error seeding kernel memory')
         return kernel_id
+
+class APIMappingKernelManager(MappingKernelManager):
+    def endpoints(self):
+        '''
+        Return a list of tuples containing the method+URI and the cell source
+        '''
+        endpoints = {}
+        for cell_source in self._source:
+            matched = self.api_indicator.match(cell_source)
+            if matched is not None:
+                uri = matched.group(2)
+                verb = matched.group(1)
+                if uri not in endpoints:
+                    endpoints[uri] = {}
+                if verb not in endpoints[uri]:
+                    endpoints[uri][verb] = cell_source
+
+        return endpoints
+
+    def sorted_endpoints(self):
+        '''Sorts the endpoints dictionary to be a list of endpoint string in order
+        from most specific to least specific.
+        '''
+        return sorted(self.endpoints(), key=self._first_path_param_index, reverse=True)
+
+    def _first_path_param_index(self, endpoint):
+        '''Returns the index to the first path parameter for the endpoint. The
+        index is not the string index, but rather where it is within the path.
+        For example:
+            first_path_param_index('/foo/:bar') # returns 1
+            first_path_param_index('/foo/quo/:bar') # return 2
+            first_path_param_index('/foo/quo/bar') # return sys.maxsize
+        '''
+        index = sys.maxsize
+        if endpoint.find(':') >= 0:
+            index = endpoint.count('/', 0, endpoint.find(':')) - 1
+        return index
+
+    @property
+    def kernelspec(self):
+        '''
+        Gets the kernel spec name required to run the notebook. Returns
+        None if no notebook exists.
+        '''
+        if hasattr(self, '_kernelspec'):
+            return self._kernelspec
+
+        if self.notebook:
+            self._kernelspec = self.notebook['metadata']['kernelspec']['name']
+        else:
+            self._kernelspec = None
+
+        return self._kernelspec
+
+    @property
+    def source(self):
+        '''
+        Gets the code source of the notebook in cell order. Returns None if no
+        notebook exists.
+        '''
+        if hasattr(self, '_source'):
+            return self._source
+
+        if self.notebook:
+            self._source = [
+                cell['source'] for cell in self.notebook.cells
+                if cell['cell_type'] == 'code'
+            ]
+        else:
+            self._source = None
+
+        return self._source
+
+    def start_kernel(self, *args, **kwargs):
+        '''
+        Starts a kernel and then optionally executes a list of code cells on it
+        before returning its ID.
+        '''
+        if self.kernelspec == 'r':
```

Comment:
Should probably refactor this into a little dictionary with a None entry for the default case.